lib/OpenQA/Schema/ResultSet/Jobs.pm

```diff
@@ -541,29 +444,15 @@ sub mark_job_linked {
 
     my $referer = Mojo::URL->new($referer_url)->host;
     my $app     = OpenQA::App->singleton;
-    if ($referer && grep { $referer eq $_ } @{$app->config->{global}->{recognized_referers}}) {
-        my $job = $self->find({id => $jobid});
-        return unless $job;
-        my $found    = 0;
-        my $comments = $job->comments;
-        while (my $comment = $comments->next) {
-            if (($comment->label // '') eq 'linked') {
-                $found = 1;
-                last;
-            }
-        }
-        unless ($found) {
-            my $user = $self->result_source->schema->resultset('Users')->search({username => 'system'})->first;
-            $comments->create(
-                {
-                    text    => "label:linked Job mentioned in $referer_url",
-                    user_id => $user->id
-                });
-        }
-    }
-    elsif ($referer) {
-        log_debug("Unrecognized referer '$referer'");
-    }
+    return undef unless $referer;
+    return log_debug("Unrecognized referer '$referer'")
+      unless grep { $referer eq $_ } @{$app->config->{global}->{recognized_referers}};
+    my $job = $self->find({id => $jobid});
+    return unless $job;
+    my $comments = $job->comments;
+    return undef if grep { $_->label // '' eq 'linked' } $comments->all;
```

Comment:
A `grep` might be slower than a loop with `last` because grep iterates over all items. `List::Util::first()` might be better here.

Reply:
The `label` function is a bit costly so I guess using `List::Util::first` is a good idea.

Reply:
oh, and `->all()` will instantiate all comments, even when using `List::Util::first(), so maybe going back to iterating via `next` is even better.

Reply:
ok. changed to https://metacpan.org/pod/List::Util#first now. Unless you show me a nicer one-line way or some performance figures I would prefer to keep the array of comment references. In python it's even expected that list operations are faster than iterating over individual members.

Reply:
Python is not perl, and even in python it would be slower to instantiate all objects and iterate over that list instead of using an iterator, in case it's expected that the loop will be exited before the end.

If the comment objects include the DateTime fields here and there are many comments, that would be unfortunate.

Reply:
But it's only an array of *references* at this point, isn't it?

Reply:
Yes, but the array of references is an array of fully instantiated objects, which include *all* columns of the comments table, including two extra costly DateTime columns.
When iterating via `next`, each object will be instantiated in the loop.
I'm not sure yet when this code is executed, so it might not make much difference, and probably there are usually not that many comments per job, but I would use an iterator.
I don't insist, I give up :)

Reply:
I'd opt for going back to the good old loop. It is as easy to understand as the new code, just a little bit longer. Saving a few lines here is likely not worth anything (even if the performance impact is very slight it is more than anything).

Reply:
If Perl is that disappointing and saving full objects in the array then, yes, it could be less efficient. We might need to change that. But reg. "not worth anything" I did it also to fix the patch code coverage check :)

Reply:
perl is not disappointing here.
`next()` will retrieve row by row via DB cursor, and then in each iteration another object is instantiated.
That is what `next()` is for - to avoid fetching data from DB (to avoid having all data in an array, or to exit the loop before the cursor is finished).
If you call `all()` then perl/DBIC does what it is **asked** and fetches **all** items from the database. And of course it has to instantiate the objects then. When should it instantiate the objects otherwise? The data must be saved somewhere.
If you want to iterate with a `for` loop or a `map` over data, then all the data must be there before the first loop iteration.

Reply:
Right, now I get it. Thank you

Reply:
Using `return undef unless $comments->find({text => {like => 'label:linked%'}})` for comment label lookup